controllers/topology/internal/extensions/patches/api/interface.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package api contains the API definition for the patch engine.
+package api
+
+import (
+	"context"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+// Generator defines a component that can generate patches for ClusterClass templates.
+// NOTE: We are introducing this interface as a decoupling layer between the patch engine and the concrete components
+// responsible for generating patches, because we aim to provide support for external patches in a future iteration.
+// We also assume that this interface and all the related types will be moved in a separated (versioned) package thus
+// providing a versioned contract between Cluster API and the components implementing external patch extensions.
+type Generator interface {
+	// Generate generates patches for templates.
+	// GenerateRequest contains templates and the corresponding variables.
+	// GenerateResponse contains the patches which should be applied to the templates of the GenerateRequest.
+	Generate(ctx context.Context, request *GenerateRequest) (*GenerateResponse, error)
+}
+
+// GenerateRequest defines the input for a Generate request.
+type GenerateRequest struct {
+	// Variables is a name/value map containing variables.
+	Variables map[string]apiextensionsv1.JSON
+
+	// Items contains the list of templates to generate patches for.
+	Items []*GenerateRequestTemplate
+}
+
+// GenerateRequestTemplate defines one of the ClusterClass templates to generate patches for.
+type GenerateRequestTemplate struct {
+	// TemplateID identifies a template to generate patches for;
+	// the same TemplateID must be used when specifying to which template a generated patch should be applied to.
+	TemplateID TemplateID
```

Comment:
```suggestion
	// TemplateID identifies a template to generate patches for;
	// the same TemplateID must be used when specifying to which template a generated patch should be applied to.
	Template Template
```

Reply:
It does feel that these should be called References, given that's almost like a custom definition of a reference, but it's definitely not an identifier like an object's uid

Reply:
Renamed to TemplateRef